File: app/src/androidTestGoogleplay/java/org/tasks/gtasks/GoogleTaskSynchronizerTest.java
package org.tasks.gtasks;

import static com.natpryce.makeiteasy.MakeItEasy.with;
import static com.todoroo.astrid.data.Task.HIDE_UNTIL_DUE;
import static com.todoroo.astrid.data.Task.HIDE_UNTIL_DUE_TIME;
import static junit.framework.Assert.assertEquals;
import static org.tasks.gtasks.GoogleTaskSynchronizer.mergeDates;
import static org.tasks.makers.TaskMaker.DUE_DATE;
import static org.tasks.makers.TaskMaker.DUE_TIME;
import static org.tasks.makers.TaskMaker.HIDE_TYPE;
import static org.tasks.makers.TaskMaker.newTask;

import android.support.test.runner.AndroidJUnit4;
import com.todoroo.astrid.data.Task;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.tasks.time.DateTime;

@RunWith(AndroidJUnit4.class)
public class GoogleTaskSynchronizerTest {

  @Test
  public void testMergeDate() {
    Task local = newTask(with(DUE_DATE, new DateTime(2016, 3, 12)));

    mergeDates(newTask(with(DUE_DATE, new DateTime(2016, 3, 11))).getDueDate(), local);

    assertEquals(new DateTime(2016, 3, 11, 12, 0).getMillis(), local.getDueDate().longValue());
  }

  @Test
  public void testMergeTime() {
    Task local = newTask(with(DUE_TIME, new DateTime(2016, 3, 11, 13, 30)));

    mergeDates(newTask(with(DUE_DATE, new DateTime(2016, 3, 11))).getDueDate(), local);

    assertEquals(new DateTime(2016, 3, 11, 13, 30, 1).getMillis(), local.getDueDate().longValue());
  }

  @Test
  public void testDueDateAdjustHideBackwards() {
    Task local =
        newTask(with(DUE_DATE, new DateTime(2016, 3, 12)), with(HIDE_TYPE, HIDE_UNTIL_DUE));

    mergeDates(newTask(with(DUE_DATE, new DateTime(2016, 3, 11))).getDueDate(), local);

    assertEquals(new DateTime(2016, 3, 11).getMillis(), local.getHideUntil().longValue());
  }

  @Test
  public void testDueDateAdjustHideForwards() {
    Task local =
        newTask(with(DUE_DATE, new DateTime(2016, 3, 12)), with(HIDE_TYPE, HIDE_UNTIL_DUE));

    mergeDates(newTask(with(DUE_DATE, new DateTime(2016, 3, 14))).getDueDate(), local);

    assertEquals(new DateTime(2016, 3, 14).getMillis(), local.getHideUntil().longValue());
  }

  @Test
  public void testDueTimeAdjustHideBackwards() {
    Task local =
        newTask(
            with(DUE_TIME, new DateTime(2016, 3, 12, 13, 30)),
            with(HIDE_TYPE, HIDE_UNTIL_DUE_TIME));

    mergeDates(newTask(with(DUE_DATE, new DateTime(2016, 3, 11))).getDueDate(), local);

    assertEquals(
        new DateTime(2016, 3, 11, 13, 30, 1).getMillis(), local.getHideUntil().longValue());
  }

  @Test
  public void testDueTimeAdjustTimeForwards() {
    Task local =
        newTask(
            with(DUE_TIME, new DateTime(2016, 3, 12, 13, 30)),
            with(HIDE_TYPE, HIDE_UNTIL_DUE_TIME));

    mergeDates(newTask(with(DUE_DATE, new DateTime(2016, 3, 14))).getDueDate(), local);

    assertEquals(
        new DateTime(2016, 3, 14, 13, 30, 1).getMillis(), local.getHideUntil().longValue());
  }

  @Test
  public void testDueDateClearHide() {
    Task local =
        newTask(with(DUE_DATE, new DateTime(2016, 3, 12)), with(HIDE_TYPE, HIDE_UNTIL_DUE));

    mergeDates(newTask().getDueDate(), local);

    assertEquals(0, local.getHideUntil().longValue());
  }

  @Test
  public void testDueTimeClearHide() {
    Task local =
        newTask(
            with(DUE_TIME, new DateTime(2016, 3, 12, 13, 30)),
            with(HIDE_TYPE, HIDE_UNTIL_DUE_TIME));

    mergeDates(newTask().getDueDate(), local);

    assertEquals(0, local.getHideUntil().longValue());
  }
}

File: app/src/androidTest/java/org/tasks/gtasks/GoogleTaskSynchronizerTest.kt
package org.tasks.gtasks

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.natpryce.makeiteasy.MakeItEasy.with
import org.tasks.data.entity.Task
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test
import org.junit.runner.RunWith
import org.tasks.makers.TaskMaker.DUE_DATE
import org.tasks.makers.TaskMaker.DUE_TIME
import org.tasks.makers.TaskMaker.HIDE_TYPE
import org.tasks.makers.TaskMaker.newTask
import org.tasks.time.DateTime

@RunWith(AndroidJUnit4::class)
class GoogleTaskSynchronizerTest {
    @Test
    fun testMergeDate() {
        val local = newTask(with(DUE_DATE, DateTime(2016, 3, 12)))
        GoogleTaskSynchronizer.mergeDates(newTask(with(DUE_DATE, DateTime(2016, 3, 11))).dueDate, local)
        assertEquals(DateTime(2016, 3, 11, 12, 0).millis, local.dueDate)
    }

    @Test
    fun testMergeTime() {
        val local = newTask(with(DUE_TIME, DateTime(2016, 3, 11, 13, 30)))
        GoogleTaskSynchronizer.mergeDates(newTask(with(DUE_DATE, DateTime(2016, 3, 11))).dueDate, local)
        assertEquals(DateTime(2016, 3, 11, 13, 30, 1).millis, local.dueDate)
    }

    @Test
    fun testDueDateAdjustHideBackwards() {
        val local = newTask(with(DUE_DATE, DateTime(2016, 3, 12)), with(HIDE_TYPE, Task.HIDE_UNTIL_DUE))
        GoogleTaskSynchronizer.mergeDates(newTask(with(DUE_DATE, DateTime(2016, 3, 11))).dueDate, local)
        assertEquals(DateTime(2016, 3, 11).millis, local.hideUntil)
    }

    @Test
    fun testDueDateAdjustHideForwards() {
        val local = newTask(with(DUE_DATE, DateTime(2016, 3, 12)), with(HIDE_TYPE, Task.HIDE_UNTIL_DUE))
        GoogleTaskSynchronizer.mergeDates(newTask(with(DUE_DATE, DateTime(2016, 3, 14))).dueDate, local)
        assertEquals(DateTime(2016, 3, 14).millis, local.hideUntil)
    }

    @Test
    fun testDueTimeAdjustHideBackwards() {
        val local = newTask(
                with(DUE_TIME, DateTime(2016, 3, 12, 13, 30)),
                with(HIDE_TYPE, Task.HIDE_UNTIL_DUE_TIME))
        GoogleTaskSynchronizer.mergeDates(newTask(with(DUE_DATE, DateTime(2016, 3, 11))).dueDate, local)
        assertEquals(
                DateTime(2016, 3, 11, 13, 30, 1).millis, local.hideUntil)
    }

    @Test
    fun testDueTimeAdjustTimeForwards() {
        val local = newTask(
                with(DUE_TIME, DateTime(2016, 3, 12, 13, 30)),
                with(HIDE_TYPE, Task.HIDE_UNTIL_DUE_TIME))
        GoogleTaskSynchronizer.mergeDates(newTask(with(DUE_DATE, DateTime(2016, 3, 14))).dueDate, local)
        assertEquals(
                DateTime(2016, 3, 14, 13, 30, 1).millis, local.hideUntil)
    }

    @Test
    fun testDueDateClearHide() {
        val local = newTask(with(DUE_DATE, DateTime(2016, 3, 12)), with(HIDE_TYPE, Task.HIDE_UNTIL_DUE))
        GoogleTaskSynchronizer.mergeDates(newTask().dueDate, local)
        assertEquals(0L, local.hideUntil)
    }

    @Test
    fun testDueTimeClearHide() {
        val local = newTask(
                with(DUE_TIME, DateTime(2016, 3, 12, 13, 30)),
                with(HIDE_TYPE, Task.HIDE_UNTIL_DUE_TIME))
        GoogleTaskSynchronizer.mergeDates(newTask().dueDate, local)
        assertEquals(0L, local.hideUntil)
    }

    @Test
    fun truncateValue() {
        assertEquals("1234567", GoogleTaskSynchronizer.truncate("12345678", 7))
    }

    @Test
    fun dontTruncateMax() {
        assertEquals("1234567", GoogleTaskSynchronizer.truncate("1234567", 7))
    }

    @Test
    fun dontTruncateShortValue() {
        assertEquals("12345", GoogleTaskSynchronizer.truncate("12345", 7))
    }

    @Test
    fun dontTruncateNull() {
        assertNull(GoogleTaskSynchronizer.truncate(null, 7))
    }

    @Test
    fun dontOverwriteTruncatedValue() {
        assertEquals("123456789", GoogleTaskSynchronizer.getTruncatedValue("123456789", "1234567", 7))
    }

    @Test
    fun overwriteTruncatedValueWithShortenedValue() {
        assertEquals("12345", GoogleTaskSynchronizer.getTruncatedValue("123456789", "12345", 7))
    }

    @Test
    fun overwriteTruncatedValueWithNullValue() {
        assertNull(GoogleTaskSynchronizer.getTruncatedValue("123456789", null, 7))
    }

    @Test
    fun overwriteNullValueWithTruncatedValue() {
        assertEquals("1234567", GoogleTaskSynchronizer.getTruncatedValue(null, "1234567", 7))
    }
}